lib/Ebook.php

```diff
@@ -797,23 +1383,318 @@ private function NullIfEmpty($elements): ?string{
+			Db::Query('
+				INSERT into EbookTags (EbookId, TagId)
+				values (?,
+				        ?)
+			', [$this->EbookId, $tag->TagId]);
 		}
+	}
+
+	private function DeleteLocSubjects(): void{
+		Db::Query('
+			DELETE from EbookLocSubjects
+			where
+			EbookId = ?
+		', [$this->EbookId]
+		);
+	}
+
+	private function InsertLocSubjects(): void{
+		foreach($this->LocSubjects as $locSubject){
+			Db::Query('
+				INSERT into EbookLocSubjects (EbookId, LocSubjectId)
+				values (?,
+				        ?)
+			', [$this->EbookId, $locSubject->LocSubjectId]);
+		}
+	}
+
+	private function DeleteCollectionMemberships(): void{
+		Db::Query('
+			DELETE from CollectionEbooks
+			where
+			EbookId = ?
+		', [$this->EbookId]
+		);
+	}
+
+	private function InsertCollectionMemberships(): void{
+		foreach($this->CollectionMemberships as $collectionMembership){
+			$collectionMembership->EbookId = $this->EbookId;
+			$collectionMembership->CollectionId = $collectionMembership->Collection->CollectionId;
+
+			Db::Query('
+				INSERT into CollectionEbooks (EbookId, CollectionId, SequenceNumber)
+				values (?,
+					?,
+				        ?)
+			', [$collectionMembership->EbookId, $collectionMembership->CollectionId, $collectionMembership->SequenceNumber]);
+
+			$collectionMembership->CollectionEbookId = Db::GetLastInsertedId();
+		}
+	}
 
-		return false;
+	private function DeleteGitCommits(): void{
+		Db::Query('
+			DELETE from GitCommits
+			where
+			EbookId = ?
+		', [$this->EbookId]
+		);
+	}
+
+	private function InsertGitCommits(): void{
+		foreach($this->GitCommits as $commit){
+			$commit->EbookId = $this->EbookId;
+
+			Db::Query('
+				INSERT into GitCommits (EbookId, Created, Message, Hash)
+				values (?,
+					?,
+					?,
+				        ?)
+			', [$commit->EbookId, $commit->Created, $commit->Message, $commit->Hash]);
+
+			$commit->GitCommitId = Db::GetLastInsertedId();
+		}
+	}
+
+	private function DeleteSources(): void{
+		Db::Query('
+			DELETE from EbookSources
+			where
+			EbookId = ?
+		', [$this->EbookId]
+		);
+	}
+
+	private function InsertSources(): void{
+		foreach($this->Sources as $source){
+			$source->EbookId = $this->EbookId;
+
+			Db::Query('
+				INSERT into EbookSources (EbookId, Type, Url)
+				values (?,
+					?,
+				        ?)
+			', [$source->EbookId, $source->Type->value, $source->Url]);
+
+			$source->EbookSourceId = Db::GetLastInsertedId();
+		}
+	}
+
+	private function DeleteContributors(): void{
+		Db::Query('
+			DELETE from Contributors
+			where
+			EbookId = ?
+		', [$this->EbookId]
+		);
+	}
+
+	private function InsertContributors(): void{
+		$allContributors = array_merge($this->Authors, $this->Illustrators, $this->Translators, $this->Contributors);
+		foreach($allContributors as $sortOrder => $contributor){
+			$contributor->EbookId = $this->EbookId;
+			$contributor->SortOrder = $sortOrder;
+
+			Db::Query('
+				INSERT into Contributors (EbookId, Name, UrlName, SortName, WikipediaUrl, MarcRole, FullName,
```

Comment:
Again, should be `Contributor::Create()`.

Reply:
Done.